### PR TITLE
Add PixelUnshuffle

### PR DESCRIPTION
**Pull request by jbschlosser**
Summary: Adds an implementation of `torch.nn.PixelUnshuffle` as the inverse operation of `torch.nn.PixelShuffle`. This addresses https://github.com/pytorch/pytorch/issues/2456

Test Plan:
```
# Unit tests.
python test/test_nn.py TestNN.test_pixel_shuffle_unshuffle

# Module test.
python test/test_nn.py TestNN.test_PixelUnshuffle

# C++ API tests.
build/bin/test_api

# C++ / python parity tests.
python test/test_cpp_api_parity.py

# JIT test.
python test/test_jit.py TestJitGeneratedFunctional.test_nn_pixel_unshuffle

# Override tests.
python test/test_overrides.py

# Type hint tests.
python test/test_type_hints.py
```

Differential Revision: D25401439

Screenshots of rendered docs:
Screen Shot 2020-12-18 at 12 19 05 PM
Screen Shot 2020-12-18 at 12 19 26 PM
Screen Shot 2020-12-18 at 12 19 34 PM
Screen Shot 2020-12-22 at 12 51 36 PM
Screen Shot 2020-12-22 at 12 51 44 PM

### CODE REVIEW

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 1]
```diff
@@ -51,4 +51,49 @@ Tensor pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
...
+Tensor pixel_unshuffle(const Tensor& self, int64_t downscale_factor) {
+  TORCH_CHECK(self.dim() >= 3,
+              "pixel_unshuffle expects input to have at least 3 dimensions, but got input with ",
+              self.dim(), " dimension(s)");
+  // Format: (B1, ..., Bn), C, H, W
+  int64_t c = self.size(-3);
+  int64_t h = self.size(-2);
+  int64_t w = self.size(-1);
+  const auto NUM_NON_BATCH_DIMS = 3;
+  const auto last_batch_dim = self.sizes().end() - NUM_NON_BATCH_DIMS;
```
Comment (gchanan) [inline aten/src/ATen/native/PixelShuffle.cpp:64]: nit: I would definitely expect something called `last_batch_dim` to be an integer type, that it's only valid on self_sizes should probably be part of the name.

Comment (gchanan) [reply aten/src/ATen/native/PixelShuffle.cpp:64]: also, doesn't this point to the first non-batch dim?  I.e. it's only the last_batch_dim because you are calling it in a function that "ends" one past.  So maybe a better name is "self_sizes_batch_end" or something.

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 2]
```diff
@@ -51,4 +51,49 @@ Tensor pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
...
+  int64_t downscale_factor_squared = downscale_factor * downscale_factor;
+  int64_t oc = c * downscale_factor_squared;
+  int64_t oh = h / downscale_factor;
+  int64_t ow = w / downscale_factor;
+
+  // First, reshape to expand height dim into (oh, downscale_factor) dims and width dim into
+  // (ow, downscale_factor) dims. This allows unshuffling to be done next by permuting dims.
+  std::vector<int64_t> expanded_shape(self.sizes().begin(), last_batch_dim);
+  expanded_shape.insert(expanded_shape.end(), {c, oh, downscale_factor, ow, downscale_factor});
+  const auto input_expanded = self.reshape(expanded_shape);
```
Comment (gchanan) [inline aten/src/ATen/native/PixelShuffle.cpp:81]: "expand" suggests something particular, which is that it called "expand" (i.e. didn't allocate new memory).  Is that what you expect here?

File: torch/nn/functional.py [hunk 3]
```diff
@@ -2815,6 +2815,26 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
     torch.Size([1, 1, 12, 12])
 """)
 
+pixel_unshuffle = _add_docstr(torch.pixel_unshuffle, r"""
+pixel_unshuffle(input, downscale_factor) -> Tensor
+
+Rearranges elements in a tensor of shape :math:`(*, C, H \times r, W \times r)` to a
```
Comment (gchanan) [inline torch/nn/functional.py:2821]: do we use `*` in the docs consistently how you've used it here (i.e. correctly?).  I'm particularly asking about the case where `*` translates to no dimensions.

Comment (jbschlosser) [reply torch/nn/functional.py:2821]: Hmm good question! From the [nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) docs:

"Input: `(N, *, H_{in})`, where `*` means any number of additional dimensions and `H_{in}` = in_features"

This description indicates to me that `N` and `H_{in}` are required dimensions, and `*` can be any number of additional dimensions in between those two. However, surprisingly to me, `nn.Linear` does accept an input shape of `(H_{in},)`. It seems that the `*` functions like "0 or more of the previous" (i.e. the Kleene star). I think it is the use of the word "additional" that makes this surprising to me, but maybe this is only a problem for me.

Similarly in [nn.L1Loss](https://pytorch.org/docs/stable/generated/torch.nn.L1Loss.html):

"Input: `(N, *)` where `*` means, any number of additional dimensions"

`nn.L1Loss` does indeed accept tensors of 0 dimensions, so again the `*` functions like a Kleene star.

**TL;DR: To match what's been done in the two referenced modules, the line here should change to: `(N, *, C, H \times r, W \times r)`. Similar updates should be done in `PixelShuffle` for consistency.**

File: aten/src/ATen/native/PixelShuffle.cpp (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/PixelShuffle.cpp:64]: Yes, you're right- it points to just after the last batch dim. To be more precise, I'll change this to `self_sizes_batch_end` as suggested.

File: aten/src/ATen/native/PixelShuffle.cpp (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/PixelShuffle.cpp:81]: Good point- `expand` as in `tensor.expand()` is not what is intended here. Ideally, I want to signal that the input has been reshaped, resulting in an increased number of input dimensions with no change to the number of elements. Would `input_reshaped` work here or is there a more specific way to name this? Any suggestions on the name for the shape itself?

File: torch/nn/functional.py (hunk 3 above)
Comment (mruberry) [reply torch/nn/functional.py:2821]: We do typically use `*` for 0 or more dimensions. See, for example, the torch.svd documentation: https://pytorch.org/docs/master/generated/torch.svd.html?highlight=svd#torch.svd. We're probably inconsistent about this in our documentation. 

I could see `N*` or `*` as appropriate. I wouldn't want to use `(N, *, C...)` unless zero or more dimensions could be inserted between a required dimension `N` and a required dimension `C`.

Comment (jbschlosser) [reply torch/nn/functional.py:2821]: > I could see N* or * as appropriate. I wouldn't want to use (N, *, C...) unless zero or more dimensions could be inserted between a required dimension N and a required dimension C.

I agree with this as it matches my intuition, and I think we should make a note to eventually make this consistent across torch.nn modules (including `nn.Linear` and `nn.L1Loss`).

Comment (jbschlosser) [reply torch/nn/functional.py:2821]: As far as this PR is concerned, I will add a note to the docs indicating what * represents here, with similar phrasing to that of `torch.svd`.

File: torch/nn/functional.py [hunk 4]
```diff
@@ -2815,6 +2815,26 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
...
+tensor of shape :math:`(*, C \times r^2, H, W)`.
```
Comment (mruberry) [inline torch/nn/functional.py:2822]: Like the pixel shuffle documentation, this references `r` before it's defined. The input argument is "downscale_factor". Also, while this first statement is true, I think the highest order bit to communicate when describing this function is that it reverses the pixel shuffle operation, which can can be linked to in the first sentence. The identify can even be elaborated on in a mathematical line showing that unshuffle(shuffle(...)) is an identity. This first sentence is OK, and the reshaping performed is important, but it loses the context of what the reshape does, and there's no mention of how the values in the tensor are permuted. That is, an extremely cynical reading of this first sentence is simply that it randomly rearranges all the elements in the tensor and puts them into the new shape.

File: torch/nn/functional.py [hunk 5]
```diff
@@ -2815,6 +2815,26 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
...
+
+See :class:`~torch.nn.PixelUnshuffle` for details.
```
Comment (mruberry) [inline torch/nn/functional.py:2824]: It always seems like an interesting quirk to me that we refer to the module over the function as our source of truth, but I think this is how most of our nn documents work.

File: torch/nn/modules/pixelshuffle.py [hunk 6]
```diff
@@ -50,3 +50,45 @@ def forward(self, input: Tensor) -> Tensor:
 
     def extra_repr(self) -> str:
         return 'upscale_factor={}'.format(self.upscale_factor)
+
+
+class PixelUnshuffle(Module):
+    r"""Rearranges elements in a tensor of shape :math:`(*, C, H \times r, W \times r)`
```
Comment (mruberry) [inline torch/nn/modules/pixelshuffle.py:56]: Same comments as for above here.

File: torch/nn/modules/pixelshuffle.py [hunk 7]
```diff
@@ -50,3 +50,45 @@ def forward(self, input: Tensor) -> Tensor:
...
+    to a tensor of shape :math:`(*, C \times r^2, H, W)`. This is the inverse operation
+    of :class:`~torch.nn.PixelShuffle`.
+
+    Note that this function can take inputs with any number of batch dimensions:
```
Comment (mruberry) [inline torch/nn/modules/pixelshuffle.py:60]: I don't think we need this note and should reinforce that `*` is zero or more batch dimensions, since that pattern appears consistently in the docs.

File: torch/nn/modules/pixelshuffle.py [hunk 8]
```diff
@@ -50,3 +50,45 @@ def forward(self, input: Tensor) -> Tensor:
...
+
+    Examples::
+
+        >>> pixel_unshuffle = nn.PixelUnshuffle(3)
+        >>> input = torch.randn(1, 1, 12, 12)
+        >>> output = pixel_unshuffle(input)
+        >>> print(output.size())
+        torch.Size([1, 9, 4, 4])
+
+    .. _Real-Time Single Image and Video Super-Resolution Using an Efficient Sub-Pixel Convolutional Neural Network:
```
Comment (mruberry) [inline torch/nn/modules/pixelshuffle.py:80]: Where is this paper referenced?

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 9]
```diff
@@ -51,4 +51,49 @@ Tensor pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
   return input_permuted.reshape(final_shape);
 }
 
+
+Tensor pixel_unshuffle(const Tensor& self, int64_t downscale_factor) {
+  TORCH_CHECK(self.dim() >= 3,
```
Comment (mruberry) [inline aten/src/ATen/native/PixelShuffle.cpp:56]: Should there also be a check that downscale factor is > 0?

File: torch/nn/modules/pixelshuffle.py (hunk 8 above)
Comment (jbschlosser) [reply torch/nn/modules/pixelshuffle.py:90]: Whoops, added the reference

File: torch/nn/functional.py [hunk 10]
```diff
@@ -2815,6 +2815,25 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
     torch.Size([1, 1, 12, 12])
 """)
 
+pixel_unshuffle = _add_docstr(torch.pixel_unshuffle, r"""
+pixel_unshuffle(input, downscale_factor) -> Tensor
+
+Reverses the :class:`~torch.nn.PixelShuffle` operation by rearranging elements in a tensor of shape :math:`(*, C, H \times r, W \times r)` to a tensor of shape :math:`(*, C \times r^2, H, W)`, where r is a downscale factor.
```
Comment (glaringlee) [inline torch/nn/functional.py:2821]: This might cause lint problem. 
max number of character per line is 120.

File: torch/nn/modules/pixelshuffle.py [hunk 11]
```diff
@@ -50,3 +54,51 @@ def forward(self, input: Tensor) -> Tensor:
 
     def extra_repr(self) -> str:
         return 'upscale_factor={}'.format(self.upscale_factor)
+
+
+class PixelUnshuffle(Module):
+    r"""Reverses the :class:`~torch.nn.PixelShuffle` operation by rearranging elements in a tensor of shape :math:`(*, C, H \times r, W \times r)` to a tensor of shape :math:`(*, C \times r^2, H, W)`, where r is a downscale factor.
```
Comment (glaringlee) [inline torch/nn/modules/pixelshuffle.py:60]: same here.
This might cause lint problem.
max number of character per line is 120.

File: torch/nn/functional.py (hunk 10 above)
Comment (jbschlosser) [reply torch/nn/functional.py]: Thanks! Just setup my local dev environment and the python linter (flake8) is not functioning correctly yet :(

**Comment by mruberry** [comment]
I reviewed the test failures, most are:

```
Dec 18 08:04:54 AssertionError: The following functions are not tested for __torch_function__ support, please ensure there is an entry in the dict returned by torch._overrides.get_testing_overrides for this function or if a __torch_function__ override does not make sense, add an entry to the tuple returned by torch._overrides.get_ignored_functions.
```

Is caused because the function is missing an override entry. See @soulitzer's recent PR adding `torch.sinc` for an idea of what files need to be updated: https://github.com/pytorch/pytorch/pull/48740. The relevant line is this one:

https://github.com/pytorch/pytorch/blob/c0deb231db76dbea8a9d326401417f7d1ce96ed5/torch/overrides.py#L774

@heitorschueroff has also been interested in helping improve our documentation for adding a new operator. There's a lot to it.

A few builds are also failing due to a couple mypy issues.

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 12]
```diff
@@ -29,24 +33,81 @@ Tensor pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
...
+              self.dim(), " dimension(s)");
+  TORCH_CHECK(
+      downscale_factor > 0,
+      "pixel_unshuffle expects a positive downscale_factor, but got ",
+      downscale_factor);
+  // Format: (B1, ..., Bn), C, H, W
+  int64_t c = self.size(-3);
+  int64_t h = self.size(-2);
+  int64_t w = self.size(-1);
+  const auto NUM_NON_BATCH_DIMS = 3;
```
Comment (mruberry) [inline aten/src/ATen/native/PixelShuffle.cpp:75]: nit: constexpr

File: torch/nn/functional.py [hunk 13]
```diff
@@ -2799,7 +2799,7 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
 pixel_shuffle(input, upscale_factor) -> Tensor
 
 Rearranges elements in a tensor of shape :math:`(*, C \times r^2, H, W)` to a
-tensor of shape :math:`(*, C, H \times r, W \times r)`.
+tensor of shape :math:`(*, C, H \times r, W \times r)`, where r is an upscale factor.
```
Comment (mruberry) [inline torch/nn/functional.py:2802]: "r is the :attr:\`upscale_factor\`."

File: torch/nn/functional.py [hunk 14]
```diff
@@ -2815,6 +2815,27 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
     torch.Size([1, 1, 12, 12])
 """)
 
+pixel_unshuffle = _add_docstr(torch.pixel_unshuffle, r"""
+pixel_unshuffle(input, downscale_factor) -> Tensor
+
+Reverses the :class:`~torch.nn.PixelShuffle` operation by rearranging elements in a
+tensor of shape :math:`(*, C, H \times r, W \times r)` to a tensor of shape
+:math:`(*, C \times r^2, H, W)`, where r is a downscale factor.
```
Comment (mruberry) [inline torch/nn/functional.py:2823]: Analogous change here as above.

File: torch/nn/modules/pixelshuffle.py [hunk 15]
```diff
@@ -6,26 +6,30 @@
 
 class PixelShuffle(Module):
     r"""Rearranges elements in a tensor of shape :math:`(*, C \times r^2, H, W)`
-    to a tensor of shape :math:`(*, C, H \times r, W \times r)`.
+    to a tensor of shape :math:`(*, C, H \times r, W \times r)`, where r is an upscale factor.
```
Comment (mruberry) [inline torch/nn/modules/pixelshuffle.py:9]: Using "an upscale factor" seems OK here since the name of the input is not available

**Review by mruberry — APPROVED** [review]
LGTM! Just some minor inline comments.

File: torch/csrc/api/include/torch/nn/options/pixelshuffle.h [hunk 16]
```diff
@@ -21,6 +21,20 @@ struct TORCH_API PixelShuffleOptions {
...
 };
 
+/// Options for the `PixelUnshuffle` module.
+///
+/// Example:
+/// ```
+/// PixelUnshuffle model(PixelUnshuffleOptions(5));
+/// ```
+struct TORCH_API PixelUnshuffleOptions {
+  PixelUnshuffleOptions(int64_t downscale_factor)
```
Comment (glaringlee) [inline torch/csrc/api/include/torch/nn/options/pixelshuffle.h:31]: nit: for single arg unique constructor, we normally put explicit before, this can avoid warnings in phabricator as well.

**Review by glaringlee — COMMENTED** [review]
@jbschlosser 
This generally looks good to me now, I have one nit comment, not mandatory.

File: torch/csrc/api/include/torch/nn/options/pixelshuffle.h (hunk 16 above)
Comment (jbschlosser) [reply torch/csrc/api/include/torch/nn/options/pixelshuffle.h:31]: Thanks for catching this! I might be wrong, but it looks like the implicit behavior is desired for C++ module options. None of the option structs defined in `torch/csrc/api/include/torch/nn/options` use `explicit`, and in fact several have the `/* implicit */` comment. Further, the C++ module tests are often written to use the implicit conversion. So to follow our standards and match these, I think I should add a similar `/* implicit */` comment here as well. Thoughts on this?

Comment (glaringlee) [reply torch/csrc/api/include/torch/nn/options/pixelshuffle.h]: @jbschlosser 
The C++ lib is contributed by many external contributors. Multiple coding styles are mixed together at the moment.
I plan to standardized it a bit in the future.
Adding /*implicit*/ is fine, I will update them in one shot later.
